gateway/ln-gateway/src/rpc/rpc_server.rs

```diff
@@ -83,23 +85,60 @@ fn extract_bearer_token(request: &Request) -> Result<String, StatusCode> {
 }
 
 /// Middleware to authenticate an incoming request. Routes that are
-/// authenticated with this middleware always require a Bearer token to be
-/// supplied in the Authorization header.
-async fn auth_middleware(
+/// authenticated with this middleware always require a Bearer token
+/// with the JWT generated previously to be supplied in the Authorization
+/// header. If jwt fails, try to check authentication with password
+async fn auth_jwt_middleware_with_password_fallback(
     Extension(gateway): Extension<Arc<Gateway>>,
     request: Request,
     next: Next,
 ) -> Result<impl IntoResponse, StatusCode> {
     let token = extract_bearer_token(&request)?;
+    let auth_manager = gateway.auth_manager.lock().await;
+    if let Ok(decoded_token_data) = jsonwebtoken::decode::<crate::auth_manager::Session>(
+        &token,
+        &jsonwebtoken::DecodingKey::from_secret(auth_manager.encoding_secret.as_ref()),
+        &jsonwebtoken::Validation::default(),
+    ) {
+        let now = fedimint_core::time::now()
+            .duration_since(UNIX_EPOCH)
+            .expect("Time went backwards")
+            .as_secs();
+        if now > decoded_token_data.claims.exp {
+            return Err(StatusCode::UNAUTHORIZED);
+        }
+        return Ok(next.run(request).await);
+    }
+    //TODO remove this fallback when all the callers support JWT auth
+    // fallback to check for password
     if bcrypt::verify(token, &gateway.bcrypt_password_hash.to_string())
         .expect("Bcrypt hash is valid since we just stringified it")
     {
         return Ok(next.run(request).await);
     }
-
     Err(StatusCode::UNAUTHORIZED)
 }
 
+async fn auth_get_jwt_session(
+    Extension(gateway): Extension<Arc<Gateway>>,
+    request: Request,
+) -> Result<impl IntoResponse, AdminGatewayError> {
+    let token = extract_bearer_token(&request).map_err(|_| AdminGatewayError::Unauthorized)?;
+    if bcrypt::verify(token, &gateway.bcrypt_password_hash.to_string())
+        .expect("Bcrypt hash is valid since we just stringified it")
+    {
+        let auth_manager = gateway.auth_manager.lock().await;
+        let session = auth_manager
+            .generate_session()
+            .map_err(|_| AdminGatewayError::Unauthorized)?;
+        let token = session
+            .encode_jwt(&auth_manager.encoding_secret)
+            .map_err(|_| AdminGatewayError::Unauthorized)?;
```

Comment:
Should be encapsulated as one function in the `AuthManager`. There's no reason to generate a session without signing it imo. Just call the fn something that makes it clear that the session tokens are signed.